### DISCUSSION TITLE
PAGView.setComposition 会触发Crash

### DISCUSSION THREAD

**Discussion by AHuier** [Tencent/libpag#1802] [General] [answered]
外网出现 PAG动画播放时crash，crash发生在JNI层（Java_org_libpag_PAGPlayer_setComposition），请PAG同事帮忙咨询一下

一、具体堆栈如下：
SIGABRT:
10-26 19:39:20.850 22403 22403 F .tencent.wesin: java_vm_ext.cc:577] JNI DETECTED ERROR IN APPLICATION: fid == null
10-26 19:39:20.850 22403 22403 F .tencent.wesin: java_vm_ext.cc:577]     in call to GetLongField
10-26 19:39:20.850 22403 22403 F .tencent.wesin: java_vm_ext.cc:577]     from void org.libpag.PAGPlayer.setComposition(org.libpag.PAGComposition)
10-26 19:39:21.748 22403 26681 E PartyRoomCoreGameService: [, , 0]:onMicSeqNoUpdated ==> 
10-26 19:39:22.431 22403 23926 E PartyRoomLyricController: [, , 0]:handleZegoChorusWatchSinger -> playTim: 146540
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655] Runtime aborting...
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655] Dumping all threads without mutator lock held
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655] All threads:
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655] DALVIK THREADS (227):
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655] "main" prio=10 tid=1 Runnable
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   | group="" sCount=0 dsCount=0 flags=0 obj=0x727926a8 self=0xb4000074a2ab3c00
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   | sysTid=22403 nice=-10 cgrp=default sched=0/0 handle=0x74a420a4f8
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   | state=R schedstat=( 1202216057384 140297521751 642847 ) utm=113184 stm=7036 core=2 HZ=100
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   | stack=0x7fd235e000-0x7fd2360000 stackSize=8192KB
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   | held mutexes= "abort lock" "mutator lock"(shared held)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #00 pc 00000000004a6e34  /apex/com.android.art/lib64/libart.so (art::DumpNativeStack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, int, BacktraceMap*, char const*, art::ArtMethod*, void*, bool)+140)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #01 pc 00000000005b5f8c  /apex/com.android.art/lib64/libart.so (art::Thread::DumpStack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool, BacktraceMap*, bool) const+372)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #02 pc 00000000005d352c  /apex/com.android.art/lib64/libart.so (art::DumpCheckpoint::Run(art::Thread*)+924)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #03 pc 00000000005cd39c  /apex/com.android.art/lib64/libart.so (art::ThreadList::RunCheckpoint(art::Closure*, art::Closure*)+532)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #04 pc 00000000005cc51c  /apex/com.android.art/lib64/libart.so (art::ThreadList::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool)+1876)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #05 pc 0000000000566304  /apex/com.android.art/lib64/libart.so (art::Runtime::Abort(char const*)+1876)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #06 pc 0000000000013ab0  /system/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_3::__invoke(char const*)+80)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #07 pc 0000000000013090  /system/lib64/libbase.so (android::base::LogMessage::~LogMessage()+320)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #08 pc 000000000038a408  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::JniAbort(char const*, char const*)+2608)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #09 pc 00000000003c369c  /apex/com.android.art/lib64/libart.so (art::JNI<false>::GetLongField(_JNIEnv*, _jobject*, _jfieldID*)+1636)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #10 pc 00000000000e7cc8  /data/app/~~0iob7N22ztycDn5JfladCA==/com.tencent.wesing-3eh76wbSa8wiCp0rvZL75Q==/lib/arm64/libpag.so (Java_org_libpag_PAGPlayer_setComposition+56)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   native: #11 pc 000000000020c2f8  /data/app/~~yCjdj6GHwJSQEMXPSW7gfw==/com.tencent.wesing-MOLue6slEymoqHfVyUkdxg==/oat/arm64/base.odex (deleted) (Java_org_libpag_PAGPlayer_setComposition__Lorg_libpag_PAGComposition_2+152)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at org.libpag.PAGPlayer.setComposition(Native method)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at org.libpag.PAGView.setComposition(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.ranklist.widget.SingerRankView.c(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.ranklist.holder.PartyRoomRankListViewHolder.r(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.ranklist.holder.PartyRoomRankListViewHolder.switchToNextRankEntry(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.ranklist.PartyRoomRankListServiceImpl$rankEntryDisplayInfoObserver$1$1.invokeSuspend(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.ranklist.PartyRoomRankListServiceImpl$rankEntryDisplayInfoObserver$1$1.a(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.ranklist.PartyRoomRankListServiceImpl$rankEntryDisplayInfoObserver$1$1.invoke(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.wesing.party.base.ViewHolderLazy$valueOnMain$1.invokeSuspend(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at kotlinx.coroutines.u0.run(SourceFile:-1)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at android.os.Handler.handleCallback(Handler.java:938)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at android.os.Handler.dispatchMessage(Handler.java:99)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at android.os.Looper.loop(Looper.java:257)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at android.app.ActivityThread.main(ActivityThread.java:8220)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at java.lang.reflect.Method.invoke(Native method)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:626)
10-26 19:39:24.843 22403 22403 F .tencent.wesin: runtime.cc:655]   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1015)

二、报错堆栈符号库还原后：
❯ aarch64-linux-android-addr2line -f -C -e libpag.so 00000000000e7cc8
_JNIEnv::GetLongField(_jobject*, _jfieldID*)
/Users/libpag/Library/Android/sdk/ndk/android-ndk-r19c/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/jni.h:709

三、案例代码
    private fun demoPAG() {
        val pagView = PAGView(this)
        container?.addView(pagView)
        pagView.setScaleMode(PAGScaleMode.Stretch)
        val pagFile = PAGFile.Load(this.assets, "live_pk_angel_show.pag")
        pagView.composition = pagFile
        pagView.layoutParams = RelativeLayout.LayoutParams(346, 172)
        pagView.play()
    }

四、PAG 版本：
1. 升级前 3.3.0.232-noffmpeg 
2. 升级后 4.1.13
升级前后均会报此crash

**Comment by kevingpqi123** [comment, marked as answer]
如果升级前后都有问题，可能的原因有两个：
（1）如果 so 是通过插件下发的，确认下执行方法前是否正确加载了 so；
（2）还有一种可能是业务侧 fd 泄漏 或内存泄漏，导致 so 读取或 JNI 方法加载失败，其它业务中出现过类似的情况
